packages/phone-number-privacy/common/src/utils/key-version.ts

```diff
@@ -0,0 +1,76 @@
+import Logger from 'bunyan'
+import { Request } from 'express'
+import { Response as FetchResponse } from 'node-fetch'
+import { ErrorMessage, KEY_VERSION_HEADER, OdisRequest, WarningMessage } from '..'
+
+export function requestHasValidKeyVersion(
+  request: Request<{}, {}, OdisRequest>,
+  logger: Logger
+): boolean {
+  try {
+    getRequestKeyVersion(request, logger)
+    return true
+  } catch {
+    return false
+  }
+}
+
+export function getRequestKeyVersion(
+  request: Request<{}, {}, OdisRequest>,
+  logger: Logger
+): number | undefined {
+  const keyVersionHeader = request.headers[KEY_VERSION_HEADER]
+  if (keyVersionHeader === undefined) {
+    return undefined
+  }
+
+  const keyVersionHeaderString = keyVersionHeader.toString() // could be string[]
+
+  if (keyVersionHeaderString.trim().length === 0) {
+    return undefined
+  }
+
+  const requestedKeyVersion = Number(keyVersionHeaderString)
+
+  if (!Number.isInteger(requestedKeyVersion) || requestedKeyVersion < 0) {
+    logger.error({ keyVersionHeader }, WarningMessage.INVALID_KEY_VERSION_REQUEST)
+    throw new Error(WarningMessage.INVALID_KEY_VERSION_REQUEST)
+  }
+  return requestedKeyVersion
+}
+
+export function responseHasValidKeyVersion(
+  request: Request<{}, {}, OdisRequest>,
+  response: FetchResponse,
+  defaultRequestKeyVersion: number,
+  logger: Logger
+): boolean {
+  const responseKeyVersion = getResponseKeyVersion(response, logger)
+  const requestKeyVersion = getRequestKeyVersion(request, logger) ?? defaultRequestKeyVersion
+
+  const isValid = responseKeyVersion === requestKeyVersion
+  if (!isValid) {
+    logger.error(
+      { requestKeyVersion, responseKeyVersion },
+      ErrorMessage.INVALID_KEY_VERSION_RESPONSE
+    )
+  }
+
+  return isValid
+}
+
+export function getResponseKeyVersion(response: FetchResponse, logger: Logger): number | undefined {
+  const keyVersionHeader = response.headers.get(KEY_VERSION_HEADER)
+  if (keyVersionHeader === null || keyVersionHeader.trim().length === 0) {
+    return undefined
+  }
+
+  const responseKeyVersion = Number(keyVersionHeader)
+
+  if (!Number.isInteger(responseKeyVersion) || responseKeyVersion < 0) {
+    logger.error({ keyVersionHeader }, ErrorMessage.INVALID_KEY_VERSION_RESPONSE)
+    return undefined
```

Comment:
Yea the domains signature endpoint doesn't verify partial signatures

Reply:
right but because it's not possible with the poprf library to do so, not because we don't want to verify those partial signatures, right? Throwing an error in this case (when it will fail on the client-side later anyways) seems like a helpful error, what do you think about adding a ticket and revisiting this post merge?